runtime-library/CMakeLists.txt

```diff
@@ -0,0 +1,118 @@
+# development as it precludes important optimizations.
+option(BUILD_SHARED_LIBS "Build development shared runtime library" OFF)
+
+# When building a shared library, further customize:
+#   - Build with hidden visibility by default.
+#   - Unhide API exported functions.
+#   - Mark symbols for dllexport on Windows by default.
+if(BUILD_SHARED_LIBS)
+  message(STATUS "IREE runtime library with BUILD_SHARED_LIBS=ON is only supported for development!")
+  set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+  add_compile_definitions("IREE_API_ENABLE_VISIBILITY")
+  add_compile_definitions("IREE_API_BUILDING_LIBRARY")
+endif()
+
+# Customize defaults.
+option(IREE_BUILD_COMPILER "Disable compiler for runtime-library build" OFF)
+option(IREE_BUILD_SAMPLES "Disable samples for runtime-library build" OFF)
+
+add_subdirectory("${IREE_ROOT_DIR}" "iree_core" EXCLUDE_FROM_ALL)
+
+#-------------------------------------------------------------------------------
+# Build the ireert library
+#-------------------------------------------------------------------------------
+
+# IREE exposes its transitive objects and deps on special target properties.
+set(_RUNTIME_ROOT_TARGET "iree::runtime::impl")
+set(_RUNTIME_OBJECTS "$<REMOVE_DUPLICATES:$<GENEX_EVAL:$<TARGET_PROPERTY:${_RUNTIME_ROOT_TARGET},INTERFACE_IREE_TRANSITIVE_OBJECTS>>>")
+set(_RUNTIME_LIBS "$<REMOVE_DUPLICATES:$<GENEX_EVAL:$<TARGET_PROPERTY:${_RUNTIME_ROOT_TARGET},INTERFACE_IREE_TRANSITIVE_OBJECT_LIBS>>>")
+ # For debugging, write out objects and deps to files.
+file(GENERATE OUTPUT "lib/ireert.objects.txt" CONTENT "${_RUNTIME_OBJECTS}\n")
+file(GENERATE OUTPUT "lib/ireert.libs.txt" CONTENT "${_RUNTIME_LIBS}\n")
+
+# Build the library (shared or static).
+add_library(ireert ${_RUNTIME_OBJECTS})
+target_include_directories(ireert INTERFACE "${CMAKE_CURRENT_BINARY_DIR}/include")
+target_link_libraries(ireert PUBLIC ${_RUNTIME_LIBS})
+set_target_properties(ireert PROPERTIES
+  ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/lib"
+  LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/lib"
+  RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/bin"
+)
+
+if(BUILD_SHARED_LIBS)
+  # Import symbols from library on use.
+  target_compile_options(ireert INTERFACE "-UIREE_API_BUILDING_LIBRARY")
+  # Disallow undefined symbols in shared library mode.
+  if(NOT IREE_ENABLE_ASAN
+    AND NOT IREE_ENABLE_MSAN
+    AND NOT IREE_ENABLE_TSAN
+    AND NOT IREE_ENABLE_UBSAN)
+    target_link_options(ireert PRIVATE
+      $<$<PLATFORM_ID:Linux>:-Wl,--no-undefined>
+      $<$<PLATFORM_ID:Darwin>:-Wl,-undefined,error>
+    )
+  endif()
+endif()
+
+#-------------------------------------------------------------------------------
+# Populate headers from the source tree.
+#-------------------------------------------------------------------------------
+
+# Copy source tree headers.
+function(_copy_runtime_source_headers)
+  set(_RUNTIME_HEADER_SOURCE_DIR "${IREE_ROOT_DIR}/runtime/src")
```

Comment:
This is surprising, is this just to ensure standalone? E.g., this simulates a shallow checkout of git repo?

Reply:
Right now, the public API does not require any generated headers, but there is nothing keeping it held together that way. I figured that when/if that changes, we would extend this to also populate from generated headers. I expect that this sample may also be extended to dynamically fetch IREE as an external project, in which case, who knows where it is... my experience with language binding folks is that they work best with simple directory trees... and this is that.